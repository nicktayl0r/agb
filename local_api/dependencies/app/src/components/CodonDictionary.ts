export const CODON_DICTIONARY: { [index: string]: string } = {
  UUU: "Phenylalanine",
  UUC: "Phenylalanine",
  UUA: "Leucine",
  UUG: "Leucine",
  CUU: "Leucine",
  CUC: "Leucine",
  CUA: "Leucine",
  CUG: "Leucine",
  AUU: "Isoleucine",
  AUC: "Isoleucine",
  AUA: "Isoleucine",
  AUG: "Methionine",
  GUU: "Valine",
  GUC: "Valine",
  GUA: "Valine",
  GUG: "Valine",
  UCU: "Serine",
  UCC: "Serine",
  UCA: "Serine",
  UCG: "Serine",
  CCU: "Proline",
  CCC: "Proline",
  CCA: "Proline",
  CCG: "Proline",
  ACU: "Threonine",
  ACC: "Threonine",
  ACA: "Threonine",
  ACG: "Threonine",
  GCU: "Alanine",
  GCC: "Alanine",
  GCA: "Alanine",
  GCG: "Alanine",
  UAU: "Tyrosine",
  UAC: "Tyrosine",
  UAA: "Stop",
  UAG: "Stop",
  CAU: "Histidine",
  CAC: "Histidine",
  CAA: "Glutamine",
  CAG: "Glutamine",
  AAU: "Asparagine",
  AAC: "Asparagine",
  AAA: "Lysine",
  AAG: "Lysine",
  GAU: "Aspartic acid",
  GAC: "Aspartic acid",
  GAA: "Glutamic acid",
  GAG: "Glutamic acid",
  UGU: "Cysteine",
  UGC: "Cysteine",
  UGA: "Stop_UGA",
  UGG: "Tryptophan",
  CGU: "Arginine",
  CGC: "Arginine",
  CGA: "Arginine",
  CGG: "Arginine",
  AGU: "Serine",
  AGC: "Serine",
  AGA: "Arginine",
  AGG: "Arginine",
  GGU: "Glycine",
  GGC: "Glycine",
  GGA: "Glycine",
  GGG: "Glycine"
};
