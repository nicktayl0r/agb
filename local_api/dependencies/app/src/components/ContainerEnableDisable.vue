<template>
  <fieldset :id="id" v-show="visibleStore" :disabled="disabled" :class="[ 'no-border', AddRemoveClasses]" ><slot/></fieldset>
</template>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";
import WidgetEnabled from "../mixins/WidgetEnabled";

@Component({ mixins: [WidgetEnabled] })
export default class ContainerEnableDisable extends Widget {}
</script>
