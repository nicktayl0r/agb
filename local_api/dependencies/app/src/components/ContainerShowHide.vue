<template>
  <div :id="id" v-show="visibleStore" :class="['container', AddRemoveClasses]"><slot/></div>
</template>

<style>
.container {
  width: 100%;
}
</style>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";

@Component
export default class ContainerShowHide extends Widget {}
</script>
