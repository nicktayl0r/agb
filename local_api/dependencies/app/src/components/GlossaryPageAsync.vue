<template>
  <component
    :is="page"
    :projectIsReady="projectLoaded"
    :pageID="glossaryPageID"
    class="glossaryPage"
  />
</template>

<script lang="ts">
import { AxiosResponse } from "axios";
import { AxiosRetryInstance } from "@/helpers/requestHelpers";
import { Component, Prop } from "vue-property-decorator";


import Page from "./Page";
import { logMessage } from "@/helpers/debugHelpers";
import { tryCatch } from "@/helpers/projectHelpers";

import { getEnvPageURL } from "@/helpers/envHelpers";
import { readProjectDataExists } from "@/store/modules/appData";
import { readModulesState } from "@/store/modules/userData";

@Component
export default class GlossaryPageAsync extends Page {
  page = Page.extend({ template: "<div>loading...</div>" });

  @Prop({ default: false })
  projectIsReady: boolean;

  @Prop()
  pageID: string;

  projectNowReady = false;

  get projectLoaded() {
    logMessage("glossary projectLoaded computed property detects change");
    const storeProject = readProjectDataExists(this.$store);
    if (!this.projectNowReady && storeProject) {
      this.projectNowReady = true;
      const pageID = this.glossaryPageID; // this will call loadPageIfValid
    }
    return storeProject;
  }

  get glossaryPageID(): string {
    const modulesState = readModulesState(this.$store);
    const pageID = modulesState.currentGlossaryPageID;
    this.loadPageIfValid(pageID);
    return pageID;
  }

  loadPageIfValid(pageID: string) {
    if (pageID) {
      this.getPageAsync(pageID);
    }
  }

  getPageAsync(pageID: string) {
    logMessage("attempting to async load glossary page", pageID);
    AxiosRetryInstance.get(getEnvPageURL(pageID), {
      responseType: "text"
    })
      .then(response => tryCatch(() => this.setPage(pageID, response)))
      .catch(error => tryCatch(() => {
        this.page = Page.extend({
          template: `<div>Couldn't load page ${pageID} because ${error}</div>`
        });
      }));
  }

  setPage(pageID: string, response: AxiosResponse) {
    const wrappedTemplate = `<div>${response.data}</div>`; // Vue wants templates to have a single parent element
    this.page = Page.extend({ template: wrappedTemplate });
  }
}
</script>

<style scoped>
.glossaryPage {
  height: 100%;
}
</style>
