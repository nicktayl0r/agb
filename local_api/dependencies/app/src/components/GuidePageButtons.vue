<template>
  <div v-show="guideOpen" :data-currentPage="guideCurrentPage">
    <widget-button id="guidePreviousPage" ref="guidePreviousPage" class="buttonBack" v-show="showPrevButton" :clickEffects="{conditionList:[ { effectsPass:[{effectData:{effectType: 'previousPage'}}] } ]}" />
    <widget-button id="guideNextPage" ref="guideNextPage" class="buttonForward" v-show="showNextButton" :class="{ buttonBgImgBlinkBlue: blinkNextPage() }" :clickEffects="{conditionList:[ { effectsPass:[{effectData:{effectType: 'nextPage'}}] } ]}" />
  </div>
</template>

<script lang="ts">
import { Component, Watch, Prop } from "vue-property-decorator";
import Vue, { VueConstructor } from "vue";
import { Maybe } from "true-myth";

import { readModulesState } from "../store/modules/userData";
import {
  getModuleFirstPageID,
  getModuleLastPageID
} from "../helpers/projectHelpers";
import Widget from "./Widget";
import WidgetButton from "./WidgetButton.vue";
import { logMessage } from "@/helpers/debugHelpers";
import {
  getNextPageID,
  isPageUnlocked,
  getNextPageUserState
} from "../helpers/userHelpers";
import { useTeacherView } from "@/helpers/widgetHelpers";

@Component({
  components: {
    WidgetButton
  }
})
export default class GuidePageButtons extends Vue {
  teacherView = false;

  guideCurrentPage = "";

  // for most widgets on pages we set their pageID on created
  // and we don't want them to navigate if the current page isn't their page
  // (if an effect is delayed, and the user navigates somewhere else)
  // since these buttons are special and exists outside of a page,
  // let's update their pageID whenever the current guide page changes in the store
  getGuideCurrentPage() {
    logMessage("GuidePageButtons.getGuideCurrentPage()");
    const modulesState = readModulesState(this.$store);
    const prevPageButton = this.$refs.guidePreviousPage as Widget;
    const nextPageButton = this.$refs.guideNextPage as Widget;

    this.guideCurrentPage = modulesState.guideState.currentPageID;
    if (prevPageButton) prevPageButton.pageID = this.guideCurrentPage;
    else {
      logMessage(
        "GuidePageButtons.prevPageButton not available cannot set guideCurrentPage"
      );
    }
    if (nextPageButton) nextPageButton.pageID = this.guideCurrentPage;
    else {
      logMessage(
        "GuidePageButtons.nextPageButton not available cannot set guideCurrentPage"
      );
    }
  }

  created() {
    this.teacherView = useTeacherView();
  }

  mounted() {
    this.getGuideCurrentPage();
  }

  get guideOpen() {
    this.getGuideCurrentPage();
    const modulesState = readModulesState(this.$store);
    return modulesState.moduleOpen === "guide";
  }

  get showPrevButton() {
    this.getGuideCurrentPage();
    const maybePageID = getModuleFirstPageID("guide");
    if (maybePageID.isJust()) {
      const pageID = Maybe.unsafelyUnwrap(maybePageID);
      logMessage(pageID);

      const modulesState = readModulesState(this.$store);
      return modulesState.guideState.currentPageID !== pageID;
    }
    return true;
  }

  get showNextButton() {
    this.getGuideCurrentPage();
    const maybeLastPageID = getModuleLastPageID("guide");
    if (maybeLastPageID.isJust()) {
      const lastPageID = Maybe.unsafelyUnwrap(maybeLastPageID);
      const modulesState = readModulesState(this.$store);
      if (modulesState.guideState.currentPageID === lastPageID) return false;
    }

    if (this.teacherView) return true;

    const maybeNextPageID = getNextPageID();
    if (maybeNextPageID.isJust()) {
      const nextPageID = Maybe.unsafelyUnwrap(maybeNextPageID);
      return isPageUnlocked(nextPageID);
    }

    return true;
  }

  blinkNextPage(): boolean {
    const maybeNextPageID = getNextPageID();
    if (maybeNextPageID.isJust()) {
      const nextPageID = Maybe.unsafelyUnwrap(maybeNextPageID);
      const unlocked = isPageUnlocked(nextPageID);

      let visitedPageBefore = false;
      const maybePageUserState = getNextPageUserState();
      if (maybePageUserState.isJust()) {
        const pageState = Maybe.unsafelyUnwrap(maybePageUserState);
        if (pageState.visited != undefined) visitedPageBefore = pageState.visited;
      }
      const blink = !visitedPageBefore && unlocked;
      return blink;
    }
    return false;
  }
}
</script>

<style>
#guidePreviousPage {
  position: absolute;
  bottom: 1.137vh;
  left: 1.725vh;
  pointer-events: all;
}
#guidePreviousPage .bgImage,
#guideNextPage .bgImage {
  width: 6vh;
  height: 6vh;
}

#guideNextPage {
  position: absolute;
  bottom: 1.137vh;
  right: 1.725vh;
  pointer-events: all;
}
</style>