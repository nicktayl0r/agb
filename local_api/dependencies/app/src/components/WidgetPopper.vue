<template>
	<div :id="id" v-show="visibleStore" @click="clicked" class="dropShadow textBackgroundBasic popper" :class="{ noSelect:notGrapes, AddRemoveClasses}" :data-anchorId="anchorIdStore">
		<slot />
		<div class="popperArrow"></div>
	</div>
</template>

<style scoped>
.popper {
	position: absolute;
	padding: 3vh;
	display: block;
	z-index: 210;
}
.popper .popperArrow {
	width: 0;
	height: 0;
	border-style: solid;
	position: absolute;
	margin: 10px;
	border-color: inherit;
}

.popper[x-placement^="top"] {
	margin-bottom: 2vh;
}
.popper[x-placement^="top"] .popperArrow {
	border-width: 2vh 2vh 0 2vh;
	border-left-color: transparent;
	border-right-color: transparent;
	border-bottom-color: transparent;
	bottom: -2vh;
	left: calc(50% - 20px);
	margin-top: 0;
	margin-bottom: 0;
}
.popper[x-placement^="bottom"] {
	margin-top: 2vh;
}
.popper[x-placement^="bottom"] .popperArrow {
	border-width: 0 2vh 2vh 2vh;
	border-left-color: transparent;
	border-right-color: transparent;
	border-top-color: transparent;
	top: -2vh;
	left: calc(50% - 20px);
	margin-top: 0;
	margin-bottom: 0;
}
.popper[x-placement^="right"] {
	margin-left: 2vh;
}
.popper[x-placement^="right"] .popperArrow {
	border-width: 2vh 2vh 2vh 0;
	border-left-color: transparent;
	border-top-color: transparent;
	border-bottom-color: transparent;
	left: -2vh;
	top: calc(50% - 20px);
	margin-left: 0;
	margin-right: 0;
}
.popper[x-placement^="left"] {
	margin-right: 2vh;
}
.popper[x-placement^="left"] .popperArrow {
	border-width: 2vh 0 2vh 2vh;
	border-top-color: transparent;
	border-right-color: transparent;
	border-bottom-color: transparent;
	right: -2vh;
	top: calc(50% - 20px);
	margin-left: 0;
	margin-right: 0;
}
.popperClickable * {
	-webkit-touch-callout: auto;
    -webkit-user-select: auto;
    -khtml-user-select: auto;
    -moz-user-select: auto;
    -ms-user-select: auto;
	user-select: auto;
	pointer-events: auto;
}
</style>

<script lang="ts">
import Popper from "popper.js";

import { mergeDeepRight } from "ramda";
import { Maybe } from "true-myth";

import { Component, Prop } from "vue-property-decorator";
import Widget from "./Widget";
import WidgetClickable from "../mixins/WidgetClickable";
import { logMessage } from "@/helpers/debugHelpers";
import {
	runWidgetEventConditions,
	WidgetPropEntryOverride,
} from "../helpers/widgetHelpers";
import { delay } from "../helpers/timeHelpers";
import { readWidgetEntryVal } from "../store/modules/userData";
import { widgetValPrimitive } from "@/data models/widgetModels";

@Component({ mixins: [WidgetClickable] })
export default class WidgetPopper extends Widget {
	@Prop() anchorId: string;

	@Prop() delay: number; // if the popper is visible on the page, delay showing for this many seconds

	@Prop({ default: false })
	delayOnce: boolean; // if true only delay the first time the popper is shown

	@Prop() options: object; // we can pass in more JSON options

	@Prop({ default: true })
	clickToDismiss: boolean;

	referenceElement: Element | null;

	popperInstance: Popper;

	popperCreateCallback = (data: object) => {
		//logMessage("popper created", data);
	};

	popperUpdateCallback = (data: object) => {}; //logMessage("popper updated", data);

	currentAnchorId: string;

	delaying = false;

	observer: MutationObserver;

	mounted() {
		logMessage("popper mounted");
		this.currentAnchorId = this.propStoreOverride("anchorId") as string;

		createPopper(this);
		window.addEventListener("resize", this.reposition, false);
	}

	updated() {
		if (this.popperInstance) this.reposition();
		else createPopper(this);
	}

	beforeDestroy() {
		window.removeEventListener("resize", this.reposition, false);
		if (this.observer) this.observer.disconnect();
		this.referenceElement = null;
		this.destroyPopperInstance();
	}

	destroyPopperInstance() {
		if (
			this.popperInstance &&
			// @ts-ignore popperInstance.state
			!this.popperInstance.state.isDestroyed
		)
			this.popperInstance.destroy();
	}

	clicked() {
		if (!this.$store || !this.notGrapes) return;

		logMessage(this.$props.id, "clicked");

		runWidgetEventConditions(this, this.$props.clickEffects, this.$props.id);

		if (this.$props.clickToDismiss) this.updateWidget("visible", false);
	}

	get anchorIdStore(): string {
		const storeAnchorId = this.propStoreOverride("anchorId");
		// console.log(
		// 	"start: wigetPopper.currentAnchorID:",
		// 	this.currentAnchorId,
		// 	"storeID: ",
		// 	storeAnchorId
		// );
		if (typeof storeAnchorId === "string") {
			if (this.currentAnchorId !== storeAnchorId) {
				this.currentAnchorId = storeAnchorId;
				if (this.popperInstance) {
					//@ts-ignore - the below is not defined in the type
					this.popperInstance.reference = document.getElementById(
						this.currentAnchorId
					);
					this.reposition();
				}
			}
			// console.log("end: wigetPopper.currentAnchorID:", this.currentAnchorId);
			return storeAnchorId;
		}
		return "";
	}

	prevVisible = false;

	get visibleStore() {
		// logMessage("widgetPopper override visibleStore");
		const isVisible = this.propStoreOverride("visible") as boolean;
		if (isVisible && !this.prevVisible) {
			// logMessage("widgetPopper now visible");
			this.reposition();
		}
		this.prevVisible = isVisible;
		return isVisible;
	}

	reposition() {
		this.$nextTick(() => {
			// logMessage("widgetPopper reposition next tick");
			if (this.popperInstance) this.popperInstance.scheduleUpdate();
		});
	}
}

function getBoundariesEl() {
	return document.getElementById("case-content");
}

async function createPopper(popperWidget: WidgetPopper) {
	if (!popperWidget.$el) {
		console.warn("couldn't find popper widget element in createPopper");
		return;
	}

	if (popperWidget.delaying) return;

	if (popperWidget.notGrapes) {
		const maybeDelayedAlready = readWidgetEntryVal(popperWidget.$store)(
			popperWidget.pageID,
			popperWidget.$props.id,
			"delayed"
		);

		if (
			popperWidget.delay &&
			popperWidget.delay !== 0 &&
			popperWidget.visibleStore &&
			!(
				popperWidget.$props.delayOnce &&
				maybeDelayedAlready.isJust() &&
				Maybe.unsafelyUnwrap(maybeDelayedAlready)
			) // we need to only delay once, and haven't done it yet
		) {
			const htmlEl = popperWidget.$el as HTMLElement;
			const displayVal = htmlEl.style.display;
			// instead of setting visible = false in the store, let's just set the style
			// that way if some other widget sets it to false with an effect we can respect that after the delay
			htmlEl.style.display = "none";
			popperWidget.delaying = true;
			// logMessage("delaying popper", popperWidget.delay);
			await delay(popperWidget.delay);
			popperWidget.delaying = false;

			const maybeVisible = WidgetPropEntryOverride(
				popperWidget.pageID,
				popperWidget.$props.id,
				"visible"
			);
			if (maybeVisible.isJust()) {
				const visible = maybeVisible.unsafelyUnwrap();
				if (visible) htmlEl.style.display = displayVal;
			}

			if (popperWidget.$props.delayOnce)
				popperWidget.updateWidget("delayed", true);
			// logMessage("popper done delaying", popperWidget);
		}

		const maybeAnchorElementVisible = WidgetPropEntryOverride(
			popperWidget.pageID,
			popperWidget.currentAnchorId,
			"visible"
		);

		if (maybeAnchorElementVisible.isJust()) {
			if (maybeAnchorElementVisible.unsafelyUnwrap() === false) {
				// logMessage(
				// 	"popper:",
				// 	popperWidget.$props.id,
				// 	"anchor widget:",
				// 	popperWidget.currentAnchorId,
				// 	"is not visible",
				// );
				popperWidget.updateWidget("visible", false);
				return;
			}
		}
	}

	popperWidget.referenceElement = document.getElementById(
		popperWidget.currentAnchorId
	);
	// this is a workaround for widget-radios, where the id is set on the input element, which is set to display:none
	if (
		popperWidget.referenceElement &&
		getComputedStyle(popperWidget.referenceElement).display == "none"
	) {
		popperWidget.referenceElement = popperWidget.referenceElement.parentElement;
	}
	// logMessage("popper referenceElement", popperWidget.referenceElement);

	if (popperWidget.referenceElement != null) {
		//logMessage("popping on", referenceElement);

		const arrowElement = popperWidget.$el.querySelector(".popperArrow");

		// set some defaults
		let options: Popper.PopperOptions = {
			placement: "top",
			modifiers: {
				preventOverflow: {
					boundariesElement: getBoundariesEl()!,
					escapeWithReference: false,
				},
				keepTogether: { enabled: false },
			},
		};
		// merge options from props, allowing to overwrite defaults
		if (popperWidget.$props.options)
			options = mergeDeepRight(options, popperWidget.$props.options);
		// overwrite that with stuff we require
		options = mergeDeepRight(options, {
			modifiers: {
				arrow: {
					element: arrowElement,
				},
			},
			onCreate: popperWidget.popperCreateCallback,
			onUpdate: popperWidget.popperUpdateCallback,
			removeOnDestroy: true,
		}) as Popper.PopperOptions;
		// I had to mod the arrow modifier function to get arrows to point without keepTogether,
		// which we don't want because it puts poppers outside the boundariesElement
		if (
			options.modifiers &&
			options.modifiers.keepTogether &&
			options.modifiers.keepTogether.enabled === false
		) {
			options = mergeDeepRight(options, {
				modifiers: {
					arrow: {
						fn: arrowModifierFunc,
					},
				},
			});
		}
		// console.log("popperOptions", options);

		popperWidget.popperInstance = new Popper(
			popperWidget.referenceElement,
			popperWidget.$el,
			options
		);
		//logMessage("popper instance", popperWidget.popperInstance);

		popperWidget.observer = new MutationObserver((mutations) => {
			// logMessage("referenceEl updated: ", mutations);
			//Go through the mutations and if any of the targets are not this element (therefore referenceEl), reposition the widget.
			for (const mut of mutations) {
				if (mut.target != popperWidget.$el) {
					// logMessage("reposition popper");
					popperWidget.reposition();
					break;
				}
			}
		});
		popperWidget.observer.observe(popperWidget.referenceElement, {
			childList: true,
			subtree: true,
			attributes: true,
		});
	} else popperWidget.destroyPopperInstance();
}

function arrowModifierFunc(data: any, options: any) {
	const arrowElement = options.element;

	const placement = data.placement.split("-")[0];
	const { popper, reference } = data.offsets;
	const isVertical = ["left", "right"].indexOf(placement) !== -1;

	const len = isVertical ? "height" : "width";
	const sideCapitalized = isVertical ? "Top" : "Left";
	const side = sideCapitalized.toLowerCase();
	const altSide = isVertical ? "left" : "top";
	const opSide = isVertical ? "bottom" : "right";
	const arrowElementSize = getOuterSizes(arrowElement)[len];

	// compute center of the popper
	const center = reference[side] + reference[len] / 2 - arrowElementSize / 2;

	// Compute the sideValue using the updated popper offsets
	// take popper margin in account because we don't have this info available
	//@ts-ignore
	const css = getStyleComputedProperty(data.instance.popper);
	//@ts-ignore
	const popperMarginSide = parseFloat(css[`margin${sideCapitalized}`], 10);
	//@ts-ignore
	const popperBorderSide = parseFloat(css[`border${sideCapitalized}Width`], 10);
	let sideValue =
		center - data.offsets.popper[side] - popperMarginSide - popperBorderSide;

	// prevent arrowElement from being placed not contiguously to its popper
	sideValue = Math.max(Math.min(popper[len] - arrowElementSize, sideValue), 0);

	data.arrowElement = arrowElement;
	data.offsets.arrow = {
		[side]: Math.round(sideValue),
		[altSide]: "", // make sure to unset any eventual altSide value from the DOM node
	};

	return data;
}

function getOuterSizes(element: any) {
	const window = element.ownerDocument.defaultView;
	const styles = window.getComputedStyle(element);
	const x =
		parseFloat(styles.marginTop || 0) + parseFloat(styles.marginBottom || 0);
	const y =
		parseFloat(styles.marginLeft || 0) + parseFloat(styles.marginRight || 0);
	const result = {
		width: element.offsetWidth + y,
		height: element.offsetHeight + x,
	};
	return result;
}

function getStyleComputedProperty(element: any, property: any) {
	if (element.nodeType !== 1) {
		return [];
	}
	// NOTE: 1 DOM access here
	const window = element.ownerDocument.defaultView;
	const css = window.getComputedStyle(element, null);
	return property ? css[property] : css;
}

function getClientRect(offsets: any) {
	return {
		...offsets,
		right: offsets.left + offsets.width,
		bottom: offsets.top + offsets.height,
	};
}
</script>
