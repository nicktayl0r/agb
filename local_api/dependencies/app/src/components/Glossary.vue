<template>
	<div id="glossaryParent" v-show="show">
		<div id="glossaryBG" @click="toggleShow"></div>
		<div id="glossaryBody">
			<div id="glossaryTitleBar">
				<b>Glossary</b>
				<widget-button id="glossaryCloseButton" class="buttonX white" @click.native="toggleShow" />
			</div>
			<div id="glossaryContent">
				<div id="glossaryDefinitionTitle">
					<b v-html="currentTermName"></b>
				</div>
				<div id="glossaryDefinition">
					<glossary-page-async />
				</div>
				<div id="glossaryRelatedTerms">
					<span v-if="RelatedTerms.length > 0">Related terms: </span>
					<span v-for="(relatedTerm, index) in RelatedTerms" :key="relatedTerm.pageID">
						<router-link :id="relatedTerm.pageID" :class="{'internalLink':true, glossarySelectedTerm: relatedTerm.pageID === currentTermID}" @click.native="clickedRelatedTerm(relatedTerm.pageID)" to="" :title='relatedTerm.name'>
							<slot /><span v-html="relatedTerm.name"></span>
						</router-link>
						<span v-if="index + 1 < RelatedTerms.length">, </span>

					</span>
				</div>
				<div id="glossarySidebarSearch">
					<input id="searchInput" ref="searchInput" type="text" placeholder="Search" v-model.trim="searchQuery">
					<widget-button id="glossarySearchClearButton" class="buttonX glossarySearchClear" @click.native="clearSearch" />
				</div>
				<div id="glossarySidebar">
					<div id="termsScroll" ref="termsScroll">
						<div v-for="term in Terms" :key="term.pageID" class="glossaryTermLink">
							<router-link :id="term.pageID" :class="{'internalLink':true, glossarySelectedTerm: term.pageID === currentTermID}" @click.native="clickedRelatedTerm(term.pageID)" to="" :title='term.name'>
								<slot /><span v-html="term.name"></span>
							</router-link>
						</div>
					</div>
				</div>
			</div>
		</div>
	</div>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";
import Vue from "vue";
import { Maybe } from "true-myth";

import GlossaryPageAsync from "@/components/GlossaryPageAsync.vue";
import WidgetButton from "@/components/WidgetButton.vue";
import WidgetLink from "@/components/WidgetLink.vue";
import {
	GlossaryModule,
	GlossaryTrack,
	GlossaryPage,
	GlossaryRelatedTerm,
} from "@/data models/projectModels";
import {
	getCurrentGlossaryTrack,
	getCurrentGlossaryPage,
} from "@/helpers/userHelpers";
import { readGlossary } from "@/store/modules/appData";
import {
	readModulesState,
	commitGlossaryOpenState,
} from "@/store/modules/userData";
import { commitSetModuleCurrentPage } from "@/store/modules/userData";

@Component({
	components: {
		GlossaryPageAsync,
		WidgetButton,
		WidgetLink,
	},
})
export default class Glossary extends Vue {
	searchQuery = "";

	show = false;

	toggleShow(): void {
		this.show = !this.show;
		commitGlossaryOpenState(this.$store, this.show);
	}

	showHide(val: boolean): void {
		this.show = val;
		commitGlossaryOpenState(this.$store, this.show);

		if (this.show) {
			this.$nextTick(() => {
				const termEl = document.getElementById(this.currentTermID);
				if (termEl) {
					// center the term in the parent scroll container
					const termPos = termEl.offsetTop;
					const termHalfHeight = termEl.getBoundingClientRect().height * 0.5;
					const scrollParent = this.$refs.termsScroll as Element;
					const parentHalfHeight = scrollParent.clientHeight * 0.5;
					const scrollPos = termPos + termHalfHeight - parentHalfHeight;
					scrollParent.scrollTop = scrollPos;
				}
			});
		}
	}

	hideIfOpen(): void {
		if (this.show) this.showHide(false);
	}

	clearSearch() {
		this.searchQuery = "";
	}

	get Terms(): GlossaryPage[] {
		const currentTrack = getCurrentGlossaryTrack();
		if (currentTrack && currentTrack.pages) {
			if (
				this.searchQuery &&
				this.searchQuery != "" &&
				this.searchQuery != " "
			) {
				try {
					const includesQueryRegEx = new RegExp(this.searchQuery, "i"); // case insensitive
					return currentTrack.pages.filter((page) =>
						includesQueryRegEx.test(page.name)
					);
				} catch (ex) {
					return [];
				}
			}
			return currentTrack.pages;
		}
		return [];
	}

	get RelatedTerms(): GlossaryRelatedTerm[] {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.relatedTerms)
			return currentGlossaryPage.relatedTerms;
		return [];
	}

	get currentTermID(): string {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.pageID)
			return currentGlossaryPage.pageID;
		return "";
	}

	get currentTermName(): string {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.name)
			return currentGlossaryPage.name;
		return "";
	}

	clickedRelatedTerm(termID: string) {
		if (!this.$store) return;

		commitSetModuleCurrentPage(this.$store, {
			moduleName: "glossary",
			pageID: termID,
		});
	}
}
</script>

<style>
#glossaryParent {
	display: flex;
	justify-content: center;
	align-items: center;
	position: relative;
}
#glossaryBG {
	position: absolute;
	background-color: rgba(0, 0, 0, 0.75);
	z-index: 200;
	width: 100%;
	height: 100%;
}
#glossaryBody {
	background-color: #5c4866;
	color: white;
	width: 90%;
	height: 90%;
	display: flex;
	flex-direction: column;
	position: absolute;
	top: 5%;
	left: 5%;
	z-index: 220;
	border-radius: 0.626rem;
}
#glossaryTitleBar {
	padding: 1.137vh;
}
#glossaryContent {
	background-color: #74617a;
	margin: 0 1.137vh 1.137vh 1.137vh;
	padding: 2.25vh;
	display: grid;
	grid-template-columns: auto 22.75vh;
	grid-template-rows: min-content 1fr min-content;
	flex-grow: 1;
	max-height: 100%;
	min-height: 0;
	height: 100%;
	border-radius: 0.626rem;
}
#glossaryDefinitionTitle {
	grid-column: 1 / span 1;
	grid-row: 1 / span 1;
}
#glossaryDefinition {
	grid-column: 1 / span 1;
	grid-row: 2 / span 1;
	max-height: 100%;
	overflow-y: auto;
	padding-right: 2.25vh;
}
#glossaryRelatedTerms {
	grid-column: 1 / span 1;
	grid-row: 3 / span 1;
	align-self: center;
	padding-top: 2.25vh;
}
#glossarySidebarSearch {
	grid-column: 2 / span 1;
	grid-row: 1 / span 1;
	padding-left: 2.25vh;
	padding-bottom: 0.2rem;
	border-left: 1px solid #5c4866;
	position: relative;
}
#glossarySidebar {
	grid-column: 2 / span 1;
	grid-row: 2 / span 2;
	border-left: 1px solid #5c4866;
	padding-left: 2.25vh;
	min-height: 0;
	overflow: hidden;
}
#termsScroll {
	position: relative;
	max-height: 100%;
	overflow-y: auto;
	overflow-x: hidden;
}
#glossaryCloseButton {
	position: absolute;
	right: 0;
	top: 0;
}
#searchInput {
  padding-right: 3.3vh;
}
.glossarySelectedTerm {
	color: #5c4866 !important;
	background-color: white;
	border-radius: 0.15em;
	padding: 0.25em;
	box-decoration-break: clone;
	-webkit-box-decoration-break: clone;
}
.glossarySelectedTerm span {
	position: relative;
}
.glossaryTermLink {
	margin: 0.75em 0 0.75em 0;
}
.glossarySearchClear {
	position: absolute;
	top: 0.57vh;
	right: 0.57vh;
}
.glossarySearchClear .bgImage {
	width: 0.75rem !important;
	height: 1.5rem !important;
}
</style>
