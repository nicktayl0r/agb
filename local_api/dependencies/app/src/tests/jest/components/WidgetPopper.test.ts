import { getLocalVue, getCleanStore } from "@/tests/jest/jestSetup";
const localVue = getLocalVue();
let store = getCleanStore();

import { shallowMount } from "@vue/test-utils";
import { WidgetBaseTests } from "@/tests/jest/components/WidgetBaseTests";
import { MixinWidgetClickableTests } from "@/tests/jest/mixins/MixinWidgetClickableTests";

import WidgetPopper from "@/components/WidgetPopper.vue";

describe("WidgetPopper", () => {
  beforeEach(() => {
    store = getCleanStore();
  });

  WidgetBaseTests(WidgetPopper);
  MixinWidgetClickableTests(WidgetPopper);

  it("matches html snapshot", () => {
    const wrapper = shallowMount(WidgetPopper, {
      propsData: {
        id: "testWidget"
      },
      localVue,
      store
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
