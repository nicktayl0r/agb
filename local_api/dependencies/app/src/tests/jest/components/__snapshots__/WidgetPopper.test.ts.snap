// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetPopper matches html snapshot 1`] = `
<div
  class="dropShadow textBackgroundBasic popper noSelect"
  data-anchorid=""
  id="testWidget"
>
   
  <div
    class="popperArrow"
  />
</div>
`;
